### PR TITLE
gnu.cmake: Set Fortran standard and obsolete flags

### LINKED ISSUE #5726: Audit usage of GNU Fortran extensions and cleanup if possible
https://gcc.gnu.org/onlinedocs/gfortran/Extensions-implemented-in-GNU-Fortran.html

As mentioned in https://github.com/E3SM-Project/E3SM/issues/5724, gfortran by default sets `-std=gnu` which enables these extensions. We should set explicit standard for GNU to ensure better portability across compilers. This Github issue can serve as a catch-all to understand what changes need to be done.

Additionally, there are several instances of the following in our codebase as our machine files have this extension enabled (`-fallow-invalid-boz`)
https://github.com/search?q=repo%3AE3SM-Project%2FE3SM+invalid-boz&type=code
https://gcc.gnu.org/onlinedocs/gfortran/BOZ-literal-constants.html

https://github.com/E3SM-Project/E3SM/blob/master/cime_config/machines/Depends.gnu.cmake#L16 (and gnugpu)
```
list(APPEND ALLOW_INVALID_BOZ_LIST
  eam/src/control/cam_history.F90
  elm/src/biogeochem/MEGANFactorsMod.F90
)
...
    e3sm_add_flags("${ITEM}" "-fallow-invalid-boz") # avoids build error for integer, parameter :: gen_hash_key_offset = z'000053db'
```

### PR DESCRIPTION
**Pull request by sarats**
Explicitly set a standard (f2008) to avoid gnu extensions. Remove obsolete flags.

We need to check and fix any existing code that relies on GNU specific extensions like real-valued array indices. 

Removed obsolete (big-endian, CISM) flags.

Fixes #5726 

@rljacob Assign this for nightly testing on next when the timing is right.

### CODE REVIEW

**Comment by sarats** [comment]
Please note that any bugfixes that are needed need to be assigned to relevant groups/POCs.

File: cime_config/machines/cmake_macros/gnu.cmake [hunk 1]
```diff
@@ -1,11 +1,16 @@
...
+# https://github.com/E3SM-Project/E3SM/issues/5726
+string(APPEND FFLAGS " -std=f2008")
+
 if (CMAKE_SYSTEM_PROCESSOR MATCHES "^aarch64")
   string(APPEND CFLAGS "-mcmodel=small")
   string(APPEND FFLAGS "-mcmodel=small")
 else()
   string(APPEND CFLAGS "-mcmodel=medium")
   string(APPEND FFLAGS "-mcmodel=medium")
 endif()
```
[the remark this answers is not in the document]
Comment (sarats) [reply cime_config/machines/cmake_macros/gnu.cmake]: Added whitespace. I think the aarch stuff was from when we tested on Fugaku (Japanese A64FX Arm machine)

**Comment by sarats** [comment]
From CI testing, we got issues with REAL*8 etc. So, need to see what else pops up.
```
/usr/local/packages/mpich/include/mpif.h:547:13:
        REAL*8 PMPI_WTIME, PMPI_WTICK
             1
Error: GNU Extension: Nonstandard type declaration REAL*8 at (1)
```

**Comment by rljacob** [comment]
Is there a way to just turn off the gnu extensions but don't specify a standard?  There's probably a ton of those in the code.

**Comment by sarats** [comment]
Evidently, real*8 itself is a GNU extension. Hmm...

https://fortranwiki.org/fortran/show/Real+precision
> In [Fortran 90](https://fortranwiki.org/fortran/show/Fortran+90) and later, intrinsic types such as real and integer have a kind attribute which guarantees a specific precision and/or range. real*8 and counterparts should no longer be used (Chin, Worth, and Greenough, 2006, p. 5).

**Comment by rljacob** [comment]
Depends what gnu means by "gnu extension".  There are some things that GNU itself added to Fortran.  real*8 is a F77 non-standard thing that predates GNU fortran.

**Comment by sarats** [comment]
So, I tried carving out exceptions for gptl and mct and then it was Scorpio's turn. Anyway, we won't be able to do whac-a-mole unless we find a workaround for real*8 etc.

```
Component gptl build complete with 2 warnings
Building mct with output to file /lustre/orion/cli115/proj-shared/sarat/e3sm_scratch/gnu3-mmf/bld/mct.bldlog.230712-175520
   Calling /lustre/orion/cli115/scratch/sarat/repos/e3sm-frontier/cime/CIME/build_scripts/buildlib.mct
Component mct build complete with 4182 warnings
Building spio with output to file /lustre/orion/cli115/proj-shared/sarat/e3sm_scratch/gnu4-mmf/bld/spio.bldlog.230712-175742
   Calling /lustre/orion/cli115/scratch/sarat/repos/e3sm-frontier/share/build/buildlib.spio
ERROR: /lustre/orion/cli115/scratch/sarat/repos/e3sm-frontier/share/build/buildlib.spio FAILED, cat /lustre/orion/cli115/proj-shared/sarat/e3sm_scratch/gnu4-mmf/bld/spio.bldlog.230712-175742
```

**Comment by sarats** [comment]
Even std=f95 issues errors for REAL*8. Fortran 90+ standards really wanted folks to use `kind` and hence prohibited 'TYPE*8'. https://gcc.gnu.org/onlinedocs/gfortran/Old-style-kind-specifications.html

We probably need a blacklist approach then. Among the GNU extensions, we definitely don't want real array indices but there doesn't seem to be a flag just to disable that.
https://gcc.gnu.org/onlinedocs/gfortran/Extensions-implemented-in-GNU-Fortran.html

**Comment by rljacob** [comment]
@sarats should we close this?

**Comment by rljacob** [comment]
closing.  requires more work.

**Pull request closed without merging**